iceoryx_hoofs/include/iceoryx_hoofs/internal/posix_wrapper/shared_memory_object/memory_map.hpp

```diff
@@ -46,27 +47,88 @@ enum class MemoryMapError
+
+    /// @brief PRIVATE and enforce the base address hint
+    PRIVATE_CHANGES_AND_FORCE_BASE_ADDRESS_HINT = MAP_PRIVATE | MAP_FIXED,
+};
+
+class MemoryMap;
+/// @brief The builder of a MemoryMap object
+class MemoryMapBuilder
+{
+    /// @brief The base address suggestion to which the memory should be mapped. But
+    ///        there is not guarantee that it is really mapped at this position.
+    ///        One has to verify with .getBaseAddress if the hint was accepted.
+    ///        Setting it to nullptr means no suggestion
+    IOX_BUILDER_PARAMETER(const void*, baseAddressHint, nullptr)
+
+    /// @brief The length of the memory which should be mapped
+    IOX_BUILDER_PARAMETER(uint64_t, length, 0U)
+
+    /// @brief The file descriptor which should be mapped into process space
+    IOX_BUILDER_PARAMETER(int32_t, fileDescriptor, 0)
+
+    /// @brief Defines if the memory should be mapped read only or with write access.
+    ///        A read only memory section will cause a segmentation fault when be written to.
+    IOX_BUILDER_PARAMETER(AccessMode, accessMode, AccessMode::READ_WRITE)
+
+    /// @brief Sets the flags defining how the mapped data should be handled
+    IOX_BUILDER_PARAMETER(MemoryMapFlags, flags, MemoryMapFlags::SHARE_CHANGES)
+
+    /// @brief Offset of the memory location
+    IOX_BUILDER_PARAMETER(off_t, offset, 0)
+
+  public:
+    /// @brief creates a valid MemoryMap object. If the construction failed the expected
+    ///        contains an enum value describing the error.
+    /// @return expected containing MemoryMap on success otherwise MemoryMapError
+    cxx::expected<MemoryMap, MemoryMapError> create() noexcept;
+};
+
+/// @brief C++ abstraction of mmap and munmap. When a MemoryMap object is
+///        created the configured memory is mapped into the process space until
+///        that object goes out of scope - then munmap is called and the memory
+///        region is removed from the process space.
+class MemoryMap
 {
   public:
+    /// @brief copy operations are removed since we are handling a system resource
     MemoryMap(const MemoryMap&) = delete;
     MemoryMap& operator=(const MemoryMap&) = delete;
+
+    /// @brief move constructor
+    /// @param[in] rhs the source object
     MemoryMap(MemoryMap&& rhs) noexcept;
+
+    /// @brief move assignment operator
+    /// @param[in] rhs the source object
+    /// @return reference to *this
     MemoryMap& operator=(MemoryMap&& rhs) noexcept;
 
+    /// @brief destructor, calls munmap when the underlying memory is mapped
     ~MemoryMap() noexcept;
-    void* getBaseAddress() const noexcept;
 
-    friend class DesignPattern::Creation<MemoryMap, MemoryMapError>;
+    /// @brief returns the base address, if the object was moved it returns nullptr
+    const void* getBaseAddress() const noexcept;
+
+    /// @brief returns the base address, if the object was moved it returns nullptr
+    void* getBaseAddress() noexcept;
```

Comment:
The optional would also make this API more obvious to the user.

Reply:
Here I would not use an optional as return value since a nullptr is only returned when you continue working on a moved object. What I wanted to state here is, that after move we have defined behavior.

Reply:
On the other hand, if you pass a `nullopt` to the ctor, what should this return? It's kind of counter intuitive to get a `nullptr` then. But I leave it up to you.

Reply:
The ctor of this class has no `void*`. This class is constructed with the `MemoryMapBuilder` where you can set the baseAddressHint with a pointer which can be a nullptr.